packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+            e.setHeader("Invalid expression in package: {} not allowed", type);
+        }
+    }
+
+    ast::TreePtr preTransformIf(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "`if`");
+        return original;
+    }
+
+    ast::TreePtr preTransformWhile(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "`while`");
+        return original;
+    }
+
+    ast::TreePtr postTransformBreak(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "`break`");
+        return original;
+    }
+
+    ast::TreePtr postTransformRetry(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "`retry`");
+        return original;
+    }
+
+    ast::TreePtr postTransformNext(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "`next`");
+        return original;
+    }
+
+    ast::TreePtr preTransformReturn(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "`return`");
+        return original;
+    }
+
+    ast::TreePtr preTransformRescueCase(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "`rescue case`");
+        return original;
+    }
+
+    ast::TreePtr preTransformRescue(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "`rescue`");
+        return original;
+    }
+
+    ast::TreePtr preTransformAssign(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "`=`");
+        return original;
+    }
+
+    ast::TreePtr preTransformHash(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "hash literals");
+        return original;
+    }
+
+    ast::TreePtr preTransformArray(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "array literals");
+        return original;
+    }
+
+    ast::TreePtr preTransformMethodDef(core::MutableContext ctx, ast::TreePtr original) {
+        illegalControlFlow(ctx, original->loc, "method definitions");
+        return original;
+    }
+};
+
+// Sanity checks package files, mutates arguments to export / export_methods to point to item in namespace,
+// builds up the expression injected into packages that import the package, and codegens the <PackagedMethods>  module.
+shared_ptr<PackageInfo> getPackageInfo(core::MutableContext ctx, ast::ParsedFile &package) {
+    ENFORCE(package.file.exists());
+    ENFORCE(package.file.data(ctx).sourceType == core::File::Type::Package);
+    // Assumption: Root of AST is <root> class.
+    ENFORCE(static_cast<ast::ClassDef *>(package.tree.get()) != nullptr);
+    ENFORCE(static_cast<ast::ClassDef *>(package.tree.get())->symbol == core::Symbols::root());
```

Comment:
```suggestion
    ENFORCE(ast::isa_tree<ast::ClassDef>(package.tree));
    ENFORCE(ast::cast_tree_nonnull<ast::ClassDef>(package.tree).symbol == core::Symbols::root());
```

Don't use `static_cast` on TreePtr. We have our own abstractions for manipulating TreePtr polymorphism using tags that are stored in the upper unused bits of the pointer.